---
title: "**Discipline project **"
author:
- "**SID:** 480139690"
output:
  html_document:
    theme: paper
    toc: true
    toc_float: true
    code_folding: hide
    css: style.css
    link_citations: true
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=10, fig.height=7, fig.path='Figs/',echo=TRUE, tidy=TRUE, warning=FALSE, message=FALSE)
```


# ----------------- tasks to finish -------------- #

- Provide a clear statement of the question you intend to address or the topic that you intend to focus on your multi-media discipline report. 

- What is your approach to addressing the question stated in (1) and what is the key technique in your approach (e.g. random forest, lasso, Bayesian network etc.)? Select ONE method and provide a concise technical description.

- Identify potential shortcomings or issues associated with the data analytics that you have performed and discuss a possible approach to address the issue. Here, a strategy doesn't necessarily refer to a model, but it must address the issue.

- Create and describe the interactive graphics (or shiny app) that illustrate one aspect of your report, and please provide the link to the shiny app (this can either be a web page or a GitHub link). 

## **Loaded Packages**
```{r Package import,message=FALSE,warning=FALSE}
library(GEOquery)  ## go to https://github.com/seandavi/GEOquery for installation details
library(R.utils)
library(reshape2)
library(ggplot2)
library(limma) ## to look at linear model in matrix annotation
library(dplyr)
library(ggbiplot)
library(factoextra)
library(tidyverse)
library(edgeR)
library(ggrepel)
```

# **Executive Summary**

## **Aim and Background**

In this study, we want to figure out **what is highly differentially expressed genes between Banff acute T-cell mediated rejection (TCMR) and normal kidney allograft biopsy specimens**. Once we identified heightened expression of differential genes between TCMR and normal groups, we can predict the kidney transplant rejection based on recipientsâ€™ gene characteristics. 

The RNA-seq expression values file **GSE131179_RAW** was obtained from Gene Expression Omnibus database (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE131179), which is a subset of kidney allograft recipients transplanted and followed at our center. We found data is RPKM measure instead of raw read count data.  

## **Data loading**

This dataset contains 34 adult kidney transplant recipients with 60466 genes recorded for each recipient, 16 recipients were categorized as Banff acute T-cell mediated rejection (TCMR) and 18 as normal. 

```{r Dataset import 1, message=FALSE}
clinical_outcome <-getGEO("GSE131179")
clinical_outcome<- clinical_outcome$GSE131179_series_matrix.txt.gz

# dim(clinical_outcome)
# experimentData(clinical_outcome)

rejection_status  <- clinical_outcome$characteristics_ch1.1
rejection_status <- unlist(lapply(strsplit(as.character(rejection_status), ": " ) , `[[` , 2))
table(rejection_status)
```

```{r Dataset import 1,message=FALSE}
# Note: please change this dir to point to the folder where your dataset is
datadir = "GSE131179_RAW"
# Read in the files
fileNames <- list.files(datadir)

# Read in all 34 files to make a table
gse_raw = c()
for(i in 1:length(fileNames)){
  temptable <- read.delim(file.path(datadir, fileNames[i]), header=TRUE)
  gse_raw <- cbind(gse_raw, temptable[,2])
  colnames(gse_raw)[i] <- colnames(temptable)[2]
}

# each column represent a recipients
rownames(gse_raw) = read.delim(file.path(datadir, fileNames[1]), header=TRUE)[,1]
dim(gse_raw)
```

## **Data Pre-processing and Quality Assessment**

A quality assessment has been performed on expression data. We are looking for samples whose gene expression value collection suffered from an abnormality that renders the data points obtained from these particular samples detrimental to our purpose.

### **Data Transformation**

For data exploration and visulisation, a log2 transformed has been applied to RNA sequence. As the expression values distributionn is highly left skewed, the $log_2$ trnasformation helps to normalise the distributions at the same time. Since expression values for gene can sometimes be zero, we plus one to original values before taking $log_2$ transformation. 

```{r log2}
summary(gse_raw[,1])
summary(log2(gse_raw[,1]+1))
gse = log2(gse_raw+1)
```

```{r boxplot}
p_box <- ggplot(melt(gse), aes(x=Var2, y=value)) +  
  geom_boxplot(outlier.colour="black", outlier.shape=16,
             outlier.size=0.5, notch=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs (x = "Recipients", y = "log2(expression value+1)") + theme_minimal()
p_box
```

After normalising, boxplot does not show heterogeneous transcript distributions (there is not any boxplot that looks very different from the rest), thus we will not drop any recipients sample data. We used the normalised gene expression data for later analysis. 


## **Differential gene expression analysis**

### **Key Approach: Fitted Liear Model with Empirical Bayes Method **

To identify differential genes, we perform pair-wise comparison of each gene between **TCMR** group and **Normal** group. In that case, we use fitted linear model to estimate the variability in the data. 

A hypothesis test was performed on $\beta_g$, the coefficient of each gene, where $g\in\{0,1,\cdots,60466\}$. Each test has the following form:

1. **Hypotheses:**
  - $H_0$: There is identical gene expression value between two groups. That is, $Y_{kg} = \mu + \epsilon_{kg}, g = 1,...,60466, k = 1,2$
  - $H_1$: There is at least one differencial gene between two groups. That is, $Y_{kg} = \mu + \alpha_k + \epsilon_{kg}, g = 1,...,60466, k = 1,2$
  
2. **Test statistic:**

We use **moderated t-statistic** to perform significance analysis given by:

$$\tilde{t_{gk}}=\frac{\hat{\beta_{gk}}}{u_{gk}\tilde{s_g}}$$

Under $H_0$, we have that $T\sim t_{n-2}$ where 2 is the rank.

The linear model for gene g has residual variance ${\sigma_g}^2$ with sample value ${s_g}^2$ and degrees of freedom $d_g$.

Below are details of code to implement this procedure:

```{r, fit}
groupname <- c(unlist(rep('ACR',16)),unlist(rep('Normal',18)))
design <- model.matrix(~ groupname + 0) 
contrast.matrix <- makeContrasts(groupnameACR-groupnameNormal, levels=design) 
fit <- lmFit(gse, design)
fit2 <- contrasts.fit(fit, contrast.matrix)
```

The **design matrix** includes the independent coefficient for each sample in two groups, say $X$; the **contrast matrix** create contrasts of these fitted linear models to make all pair-wise comparisons between the two groups, say $C$. Here we did not specify the weight matrix. 

**lmFit** fits a linear model using weighted least squares for each gene. This linear model is fitted to the expression data for each probe.

After using **contrast.fit** to estimate contrast for each gene, the covariance matrix of the estimated $\beta_g$ is ${\sigma_g}^2C^T(X^TX)^{-1}C$ and the $s_g$ is the square-roots of diagonal elements of $C^T(X^TX)^{-1}C$. 

```{r eBayes, warning=FALSE}
fit2 <- eBayes(fit2)  
```

The **eBayes()** uses an empirical Bayes method to moderate the standard errors of the estimated log-fold changes(shrinks standard errors that are much larger or smaller than those from other genes towards the average standard error).This results in more stable inference and improved power.The posterior values for the residual variances are given by:
$$\tilde{s_{g}}^2=\frac{d_0s_0^2}{d_0+d_g}$$
Where $dg$ is the residual degrees of freedom for the gth gene. 

```{r topTable}
tT <- topTable(fit2) 
round(tT[1:5,], 2)
```

Using coefficient $\beta_g$ and unscaled standard deviations $s_g$, we can calculate **p-value** and order genes by significance. A list of top genes differential expressed in **TCMR** group versus **Normal** group can be obtained from topTable, which use moderated t-statistic to perform significance analysis. 

The following five genes to be a potential signature for gene differentiation: 

- `ENSG00000140749.8`
- `ENSG00000274712.1`
- `ENSG00000270231.3` 
- `ENSG00000206337.9` 
- `ENSG00000237499.5`

```{r qqt}
# length(which(tT$adj.P.Val < 0.05))
# Q-Q plot of moderated t-statistics
# qqt(fit2$t[,1],df=fit2$df.residual+fit2$df.prior)
# abline(0,1)
```

### MA-plot

```{r MA_plot}
df<- topTable(fit2, number=nrow(fit2), genelist=rownames(gse))

p <- ggplot(df, aes(x = AveExpr, y = logFC))+
    geom_point(aes(colour=-log10(P.Value)), alpha=1/3, size=1) +
    scale_colour_gradient(low="blue",high="red")+
    labs(x = "Average expression",
         y = "log2 fold change",
         title = "MA-plot")+
  theme(plot.title = element_text(hjust = 0.5)) 
p
```

From MA-plot, red represent upregulated and purple represent downregulated genes. We can see most of the upregulated concentrated in high M value, which means one group has more higher gene expression value compare to another. 

```{r vocano plot}
cut_off_pvalue = 0.005
cut_off_logFC = 1

dataset = df
dataset$change = ifelse(dataset$P.Value < cut_off_pvalue & abs(dataset$logFC) >= cut_off_logFC, 
                     ifelse(dataset$logFC > cut_off_logFC ,'Up','Down'),
                     'Stable')
p<-ggplot(
  dataset, 
  aes(x = logFC, 
      y = -log10(P.Value), 
      colour=change)) +
      geom_point(alpha=0.4, size=3.5) +
      scale_color_manual(values=c("#546de5", "#d2dae2","#ff4757"))+

  # abline
  geom_vline(xintercept=c(-1,1),lty=4,col="black",lwd=0.8) +
  geom_hline(yintercept = -log10(cut_off_pvalue),lty=4,col="black",lwd=0.8) +

  # axis
  labs(x="log2(fold change)",
       y="-log10 (p-value)")+
  theme_bw()+

  # theme
  theme(plot.title = element_text(hjust = 0.5), 
        legend.position="right", 
        legend.title = element_blank()
)

dataset$label = ifelse(dataset$P.Value < cut_off_pvalue & abs(dataset$logFC) >= 4.4, dataset$ID,"")
p <- p+geom_text_repel(data = dataset, aes(x = logFC, 
                                      y = -log10(P.Value), 
                                      label = label),
                  size = 3,box.padding = unit(0.5, "lines"),
                  point.padding = unit(0.8, "lines"), 
                  segment.color = "black", 
                  show.legend = FALSE) + 
                  labs(title = 'Vocano Plot')
p
```

From Vocano plot, we can see there are much more upregulated genes than downregulated genes, which means the final gene expression value between two groups are **not in the same order of magnitude**. That might cause biased result in later analysis. 

We want to filter the weakly-expressed genes. So next we applied PCA to reduce multidimentional datasets to lower dimensions for anlysis. 

### **PCA Analysis**

```{r PCA}
gse_pca <- prcomp(t(gse))
df_toplot <- data.frame(rejection_status, 
                        pc1 = gse_pca$x[,1], pc2 = gse_pca$x[,2]  )

p1 = ggbiplot::ggscreeplot(gse_pca) + theme_bw(base_size = 16) + 
    labs(y = "Prop. of explained variance")
p2 = ggbiplot::ggscreeplot(gse_pca, type = "cev") + theme_bw(base_size = 16) + 
    labs(y = " Cumulative prop. of explained variance")
gridExtra::grid.arrange(p1, p2, ncol = 2)

```

We found 2 principles already covered nearly 90% variance of all genes. Next we use the first two principles to visulise recipients information in 2D scatter plot.

```{r }
g <- ggplot(df_toplot, aes(x = pc1, y = pc2, color = rejection_status)) + 
  geom_point() + 
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(title = "2 Priniple PCA Analysis") 
g
```

We can easily distinguish Acute rejection recipients and Normal recipients from 2 principles. 

### **Gene Clustering**

Next, we use a clustering image map (CIM) to visulise gene clustering observation between TCMR group and Normal group for top 50 important genes.

```{r CIM,fig.height=6,fig.width=6}
### making a CIM
sig50 <- rownames(fit2$coefficients[order(fit2$F,decreasing=TRUE)[1:50],-1])
temp = gse[sig50,] 
colnames(temp) = groupname
heatmap(temp,main = "Clustering image Map",cex.main=5) 
```

Using PCA analysis result, we will keep 6047 important gene (explain 90% variance) and save it as final gene expression matrix.  

```{r }
## quick filter to reduce computational time
largevar = apply(gse, 1, var)
ind = which(largevar > quantile(largevar, 0.9))
write.csv(gse[ind,], "GSE131179_expression_matrix.csv")
temp = read.csv("GSE131179_expression_matrix.csv")
X = t(temp)[2:35,]
# X = as.matrix(t(gse[sig50,]))
```

## Generate Machine Learning classifier

After knowing significant genes, we next using Machine learning models to build a classifier for predicting kidney transplant rejection based on recipientsâ€™ gene characteristics. 

```{r classifer,warning=FALSE}
y = groupname
data = cbind(t(X),y) %>% as.data.frame()

cvK = 5  # number of CV folds
cv_50acc5_knn = cv_50acc5_svm = cv_50acc5_rf = c()
cv_acc_knn = cv_acc_svm = cv_acc_rf = c()
n_sim = 50 ## number of repeats

for (i in 1:n_sim) {

  cvSets = cvTools::cvFolds(nrow(X), cvK)  # permute all the data, into 5 folds
  cv_acc_knn = cv_acc_svm = cv_acc_rf = c()
  
  for (j in 1:cvK) {
    test_id = cvSets$subsets[cvSets$which == j]
    X_test = X[test_id, ]
    X_train = X[-test_id, ]
    y_test = y[test_id]
    y_train = y[-test_id]
    
    ## KNN
    fit5 = class::knn(train = X_train, test = X_test, cl = y_train, k = 5)
    cv_acc_knn[j] = table(fit5, y_test) %>% diag %>% sum %>% `/`(length(y_test))
    
    ## SVM
    svm_res <- e1071::svm(x = X_train, y = as.factor(y_train))
    fit <- predict(svm_res, X_test)
    cv_acc_svm[j] = table(fit, y_test) %>% diag %>% sum %>% `/`(length(y_test))

    ## RandomForest
    rf_res <- randomForest::randomForest(x = X_train, y = as.factor(y_train))
    fit <- predict(rf_res, X_test)
    cv_acc_rf[j] = table(fit, y_test) %>% diag %>% sum %>% `/`(length(y_test))
  }
  cv_50acc5_knn <- append(cv_50acc5_knn, mean(cv_acc_knn))
  cv_50acc5_svm <- append(cv_50acc5_svm, mean(cv_acc_svm))
  cv_50acc5_rf <- append(cv_50acc5_rf, mean(cv_acc_rf))
} ## end for

KNN = as.data.frame(cv_50acc5_knn)
RandomForest = as.data.frame(cv_50acc5_rf)
SVM = as.data.frame(cv_50acc5_svm)

boxplot(RandomForest, main = "Mean Accuracy of Random Forest")

cv_acc_model = cbind(KNN,RandomForest,SVM) %>% dplyr::rename(KNN = 'cv_50acc5_knn', Random_Forest = 'cv_50acc5_rf', SVM = 'cv_50acc5_svm')
gather(cv_acc_model,classifire) %>%
  ggplot(aes(x = classifire, y = value, fill = classifire)) +
  geom_boxplot(show.legend = FALSE) +
  geom_point(show.legend = FALSE) +
  labs(x = "Classifire",
       y = "Accuracy",
       title = "Accuracy distribution for different classifier") +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"),
        legend.position = "none")

```

It can be seen from the accuracy boxplots, that the Random forest was the best out of the three models, with average accuracy up to $\color{skyblue}{\text{0.87}}$.

Random forest works by randomly select sample (row) and genes (column) and build decision tree. Then random forest summarise those predictor generated in subsamples see which genes is predicted most often. 












