<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Predicting the presence of T-cell mediated rejection (TCMR) based on recipients’ gene characteristics</title>
    <meta charset="utf-8" />
    <meta name="author" content="SID: 480139690" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/fc-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolis_white.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Predicting the presence of T-cell mediated rejection (TCMR) based on recipients’ gene characteristics
## Investigate highly differentially expressed genes between Banff acute T-cell mediated rejection (TCMR) and normal kidney allograft biopsy specimens
### SID: 480139690

---






class: inverse, center, middle

# Outline

---

background-image: url(https://raw.github.sydney.edu.au/xibi6298/Data-3888-Discipline-Project-2/master/presentation%20slide-figure/outline.png?token=AAABC2BRPEKJ6JEYYVCINE26X7LEO)
background-size: 400px
background-position: 50% 75%

# Outline

In this study, we want to figure out **How to predict presence of T-cell mediated rejection (TCMR) based on recipients’ gene characteristics?**. There are 60466 genes recorded for 34 recipients. Once we identified heightened expression of differential genes between TCMR and normal groups, we can predict the kidney transplant rejection with SVM Classifier. 


---

class: inverse, center, middle

# Data loading

---

background-image: url(https://raw.github.sydney.edu.au/xibi6298/Data-3888-Discipline-Project-2/master/presentation%20slide-figure/sample%20distribution.png?token=AAABC2HFYL2W7W3PJ2GDIWK6XPB3W)
background-size: 400px
background-position: 50% 65%

# Data loading

We accessed RNA-seq expression values file **GSE131179_RAW** from Gene Expression Omnibus database (NCBI), obtained from kidney allograft recipients transplanted and followed at our center. We found data has been normalised using Fragments Per Kilobase of transcript per Million (FPKM).

---

class: inverse, center, middle

# Data Pre-processing and Quality Assessment

---

# Data Pre-processing and Quality Assessment

To minimise variance difference among different samples, we normalised data using `\(log_2\)` transformation, with 1 plus in case original expression value is 0.

&lt;img src="presentation_files/figure-html/boxplot-1.png" width="720" style="display: block; margin: auto;" /&gt;

---

class: inverse, center, middle

# Differential gene expression analysis

---

# Differential gene expression analysis

**Key Approach: Contrasts comparison with Fitted Linear Model **

We perform Contrast comparison of each gene between **TCMR** group and **Normal** group. 

1. **Hypotheses:**
  - `\(H_0\)`: There is identical gene expression value between two groups. That is, `$$Y_{kg} = \mu + \epsilon_{kg}, g = 1,...,60466, k = 1,2$$`
  - `\(H_1\)`: There is at least one differencial gene between two groups. That is, `$$Y_{kg} = \mu + \alpha_k + \epsilon_{kg}, g = 1,...,60466, k = 1,2$$`
  
2. **Test statistic:**

  - We use **moderated t-statistic** to perform significance analysis given by:   `\(\tilde{t_{gk}}=\frac{\hat{\beta_{gk}}}{u_{gk}\tilde{s_g}}\)`

3. **Empirical Bayes Method**

  - An empirical Bayes method has been applied to moderate the standard errors of the estimated log-fold changes. `\(\tilde{s_{g}}^2=\frac{d_0s_0^2}{d_0+d_g}\)`

---

class: inverse, center, middle

# Differential gene expression analysis

A list of top genes differential expressed


```
##                   logFC AveExpr     t P.Value adj.P.Val      B
## ENSG00000140749.8 2.303   1.947 9.936       0         0 13.847
## ENSG00000274712.1 0.482   0.636 9.500       0         0 12.735
## ENSG00000270231.3 0.639   1.466 8.925       0         0 11.230
## ENSG00000206337.9 1.618   3.033 8.865       0         0 11.068
## ENSG00000237499.5 0.641   0.567 8.767       0         0 10.807
```

---

# Vocano Plot

&lt;img src="presentation_files/figure-html/unnamed-chunk-2-1.png" style="display: block; margin: auto;" /&gt;

---

# Shiny App &amp; Select Minimum Number of Important Genes

&lt;img src="presentation_files/figure-html/CIM-1.png" style="display: block; margin: auto;" /&gt;

---

class: inverse, center, middle

# Machine Learning classifier

---

# Machine Learning Classifier

&lt;img src="presentation_files/figure-html/unnamed-chunk-3-1.png" style="display: block; margin: auto;" /&gt;


---

# Potential Shortcomings

-  Abnormally Distributed Sample 

  - Strategy: to keep all samples value has similar distribution, we can consider remove some really abnormal samples. 

- Majority genes are upregulated genes

  - Strategy: we can check if there is rRNA or other types of RNA affect their quality.
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
